lib/router/src/lib.rs

```diff
@@ -208,43 +246,86 @@ where
     fn not_recognized() -> Self {
         Self::error(error::NotRecognized.into())
     }
-
-    fn no_capacity(capacity: usize) -> Self {
-        Self::error(error::NoCapacity(capacity).into())
-    }
 }
 
-impl<Req, Svc> Future for ResponseFuture<Req, Svc>
+impl<Req, Rec, Mk> Future for ResponseFuture<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req> + Clone,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    type Item = <LoadShed<Svc> as svc::Service<Req>>::Response;
+    type Item = <LoadShed<Mk::Value> as svc::Service<Req>>::Response;
     type Error = error::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
+                State::Inserting(ref mut rr) => {
+                    // Aquire the lock for the router cache
+                    let mut cache = match rr.cache.poll_lock() {
+                        Async::Ready(aquired) => aquired,
+                        Async::NotReady => return Ok(Async::NotReady),
+                    };
+
+                    let request = rr.request.take().expect("polled after ready");
+                    let target = rr.target.take().expect("polled after ready");
+
+                    // If the target is already cached, route the request to
+                    // the service; otherwise, try to insert it
+                    if let Some(service) = cache.access(&target) {
+                        State::Calling(Some(request), Some(service))
+                    } else {
+                        // Ensure that there is capacity for a new slot
+                        cache.at_capacity()?;
+
+                        // Make a new service for the target
+                        let make = rr.make.take().expect("polled after ready");
+                        let service = LoadShed::new(make.make(&target));
+
+                        cache.insert(target, service.clone());
+                        State::Calling(Some(request), Some(service))
+                    }
+                }
+                State::Calling(ref mut request, ref mut service) => {
+                    let mut service = service.take().expect("polled after ready");
+
+                    match service.poll_ready().map_err(error::Error::from)? {
```

Comment:
tioli: slightly clearer to only handle the exceptional case in a block; also a prime case for `assert!`:

```rust
assert!(service.poll_ready()?.is_ready(), "...");

let request = request.take().expect("polled after ready");
State::Called(service.call(request))
```